yt_dlp/extractor/tfo.py

```diff
@@ -1,48 +1,246 @@
-import json
+import urllib.parse
 
 from .common import InfoExtractor
-from ..networking import HEADRequest
-from ..utils import ExtractorError, clean_html, int_or_none
+from .uplynk import UplynkBaseIE
+from ..utils import (
+    ExtractorError,
+    clean_html,
+    int_or_none,
+    parse_iso8601,
+    str_or_none,
+    url_or_none,
+)
+from ..utils.traversal import traverse_obj
 
 
-class TFOIE(InfoExtractor):
-    _GEO_COUNTRIES = ['CA']
-    _VALID_URL = r'https?://(?:www\.)?tfo\.org/(?:en|fr)/(?:[^/]+/){2}(?P<id>\d+)'
-    _TEST = {
-        'url': 'http://www.tfo.org/en/universe/tfo-247/100463871/video-game-hackathon',
-        'md5': 'cafbe4f47a8dae0ca0159937878100d6',
+class TFOIE(UplynkBaseIE):
+    IE_NAME = 'tfo'
+    IE_DESC = 'Télévision française de l\'Ontario'
+
+    _BASE_URL = 'https://www.tfo.org'
+    _VALID_URL = r'https?://(?:www\.)?tfo\.org/(?:episode|film|regarder|titre)(?:/[\w-]+)+/(?P<id>(?:GP)?\d{6})'
+    _TESTS = [{
+        'url': 'https://www.tfo.org/regarder/bardot-la-meprise/GP701766',
+        'info_dict': {
+            'id': '02a8e473f171403184a0a8cddc36845a',
+            'ext': 'mp4',
+            'title': 'Bardot, la Méprise',
+            'age_limit': 13,
+            'alt_title': 'bardot-la-meprise',
+            'description': 'md5:16ca832101b6c3838bb61cd8fa06aa9e',
+            'display_id': 'GP701766',
```

Comment:
Unless you have reason to believe that the `display_id` values are not unique, I would just use those as the `id`. This would make download archive functionality much more efficient for this site, since yt-dlp could recognize the video as already being recorded in the archive just from the URL, instead of having to fully extract it to get the Uplynk ID.